File: Algorithms/Searching Algorithms/Binary Search/BinarySearch.java
// package com.company;

public class BinarySearch {
    public static void main(String[] args) {
        int[] arr = {-1, 2, 5,23, 34, 53, 54, 65, 76, 88, 90};
        int target = 5;
        int ans = binarySearch(arr, target);
        System.out.println(ans);


    }
     //return the index
    // return -1 if it does not exist
    static int binarySearch(int[] arr , int target){
        int start = 0;
        int end = arr.length-1;

        while(start <= end){
            //find the middle element
//            int mid = (start + end)/2 ; might be possible that (start+ end) may exceed
            int mid = start + (end - start)/2;
            if(target < arr[mid]){
                end = mid - 1;

            } else if (target > arr[mid]){
                start = mid +1;
            }else{
                //ans found
                return mid;

            }

        }
        return -1;
    }
}

// Credit to Kunal Kushwaha

File: Algorithms/Searching Algorithms/Binary Search/BinarySearch.cpp
#include<iostream>
using namespace std;

int binarySearch(int arr[], int size, int key){

    int start=0;
    int end=size-1;

    int mid = start + (end-start)/2;

    while(start<=end){

        if(arr[mid] == key){
            return mid;
        }
        //go to  right wala part
        if(key>arr[mid]){
            start=mid+1;
        }
        else //key<arr[mid]
        {
            end = mid-1;
        }

        mid = start + (end-start)/2;
    }
        return -1;
}

int main(){

    int even[6] = {2,4,6,8,12,18};
    int odd[5] = {3,8,11,14,16};

    int evenindex= binarySearch(even, 6, 20);

    cout<<" Index of 18 is "<< evenindex <<endl;

    int oddIndex = binarySearch(odd, 5, 11);
    cout<<" Index of 11 is "<< oddIndex <<endl;

    return 0;
}
